runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsPusher.java

```diff
@@ -0,0 +1,157 @@
+      metricsContainerStepMaps = new ArrayList<>();
+      start();
+    }
+    /*
+      MetricsPusher.init will be called several times in a pipeline
+      (e.g Flink calls it with each UDF with a different MetricsContainerStepMap instance).
+      Then the singleton needs to register a new metricsContainerStepMap to merge
+      */
+    addMetricsContainerStepMap(metricsContainerStepMap);
+  }
+
+  private static void configureFromPipelineOptions(PipelineOptions pipelineOptions) {
+    switch (pipelineOptions.getMetricsSink()) {
+      case "org.apache.beam.runners.core.metrics.MetricsHttpSink":
+        metricsSink = new MetricsHttpSink(pipelineOptions.getMetricsHttpSinkUrl());
+        break;
+      default:
+        metricsSink = new DummyMetricsSink();
+    }
+    period = pipelineOptions.getMetricsPushPeriod();
+  }
+
+  /**
+   * Lazily set the pipelineResult.
+   * @param pr
+   */
+  public static void setPipelineResult(PipelineResult pr){
+    pipelineResult = pr;
+  }
+  private static void start() {
+    ScheduledExecutorService scheduler =
+        Executors.newSingleThreadScheduledExecutor(
+            new ThreadFactoryBuilder()
+                .setDaemon(false)
+                .setNameFormat("MetricsPusher-thread")
+                .build());
+    scheduledFuture = scheduler
+        .scheduleAtFixedRate(new PushingThread(), 0, period,
+            TimeUnit.SECONDS);
+  }
+  private static synchronized void tearDown(){
+    if (!scheduledFuture.isCancelled()) {
+      scheduledFuture.cancel(true);
+    }
+    /*
+    it is the end of the pipeline, so set the instance to null so that it can be initialized
+    for the next pipeline runnin gin that JVM
+    */
+    instance = null;
+  }
+
+  public static void pushMetrics(){
+    try {
+      // merge metrics
+      for (MetricsContainerStepMap metricsContainerStepMap : metricsContainerStepMaps) {
+        MetricResults metricResults = asAttemptedOnlyMetricResults(metricsContainerStepMap);
+        MetricQueryResults metricQueryResults =
+            metricResults.queryMetrics(MetricsFilter.builder().build());
+        if ((Iterables.size(metricQueryResults.distributions()) != 0)
+            || (Iterables.size(metricQueryResults.gauges()) != 0)
+            || (Iterables.size(metricQueryResults.counters()) != 0)) {
+          metricsSink.writeMetrics(metricQueryResults);
+        }
+      }
+      if (pipelineResult != null) {
+        PipelineResult.State pipelineState = pipelineResult.getState();
+        if (pipelineState.isTerminal()) {
+          tearDown();
+        }
+      }
+
+    } catch (Exception e) {
+      MetricsPushException metricsPushException = new MetricsPushException(e);
+      metricsPushException.printStackTrace();
+    }
+  }
+
+  private static class PushingThread implements Runnable {
+
+    @Override
+    public void run() {
+      pushMetrics();
+    }
+  }
+
+  /**
+   * Exception related to MetricsPusher to wrap technical exceptions
```

Comment:
Just a small remark - Jenkins PreCommit_Java_MavenInstall complains on this:
_[ERROR] src/main/java/org/apache/beam/runners/core/metrics/MetricsPusher.java:[149] (javadoc) JavadocStyle: First sentence should end with a period._